feat: Create UserRepository & Implement Sign In logics

minwukim opened the pull request:
## 관련 이슈
close #11 

## 예상 리뷰 시간
60 min

## 추가 또는 변경사항
**1.  UserRepository를 구현했습니다. 구체적으로 구현한 기능은 아래와 같습니다.**
 - Spring Data JPA를 활용하여 인터페이스 구현했습니다.
 - findByNickname, findByEmail, findByPhoneNumber 를 Optional(User)를 리턴하는 식으로 구현했습니다.

**2. UserRepository의 세가지 findBy에 대한 test cases를 작성했습니다.**

**3. SignUpRequest라는 dto를 만들었습니다. 클라이언트에서 보내주는 회원가입 정보를 매핑하기 위한 dto입니다.** 
 - nickname, phoneNumber, email, firstName, lastName 등 필수적인 필드에 @NotNull 어노테이션을 붙였습니다.
 - 필수가 아닌 profileImgURL과 statusMsg에는 @Notnull 어노테이션을 붙이지 않았습니다.


**4. UserAPI 에 SignIn 기능을 구현했습니다. 로직의 절차는 아래와 같습니다:**
 - 클라이언트에서 보내주는 회원가입 정보를 SignUpRequest dto에 매핑합니다.
 - email 포맷을 검사합니다.
 - phoneNumber 중복 여부를 검사합니다.
 - email 중복여부를 검사합니다.
 - 닉네임 중복여부를 검사합니다.
 - 중복여부 검사를 마친뒤, User 인스턴스 user를 생성하고 SignIn메서드를 통해 필드값을 세팅합니다.
 - UserRepository에 user를 세이브합니다. 

**5. UserAPI에 회원가입 전 중복체크 기능을 추가하였습니다.**
 - 이메일 중복 및 유효 검사를 하는 "validate-email"를 추가하였습니다.
 - 닉네임 중복 검사를 위한 "validate-nickname"을 추가하였습니다.
 - 전화번호 중복 검사를 위한 "validate-phone-number"를 추가하였습니다.  

**7. 중복체크 기능을 구현하기 위해 Request를 매핑할 dto를 만들었습니다. 구체적으로 아래와 같습니다:**
 - EmailValidateRequest
 - NicknameValidateRequest
 - PhoneNumberValidateRequest 

## 스크린샷
<!-- 추가되거나 변경된 사항을 이미지로 남겨주세요. -->
없음

## 기타
<!-- 작업 중 있언던 것들을 자유롭게 작성합니다. -->
"자주 풀리퀘스트를 쏘면 리뷰하는데 오래 걸리기만 한다"는 생각, 그리고 "정리가 안 됐지만 일단 지르고 수정하자"라는 생각에 PR을 미루고 미루다 이제야 올립니다.

PR의 상세내용을 작성하는 지금와서 보니 제가 조급한 나머지 너무 소통에 소홀했구나 느낍니다.
큰 뭉치를 한 방에 올리는게 얼마나 팀원분 들에게 불친절한 것인지 이제서야 깨닫네요. 반성합니다.
앞으로 잘게잘게 자주자주 PR을 쏘고 팀원분들과 자주 소통하도록 노력하겠습니다.

minwukim committed:
feat: Sign in implementation

1. Implemented "UserRepository"
2. Implemented APIs for sign in, phone number repetition check, and nickname repetition check

minwukim committed:
feat: update for sign in implementation

minwukim committed:
feat: validate email with regex

minwukim committed:
test: Add test cases for UserRepository

minwukim commented:
아, 추가적으로 언급할 부분이 있습니다. 

현재 프로필 사진은 스트링으로 취급한 상태입니다. 추후 form 형식으로 받아오는 식으로 수정해야합니다.

JinuCheon commented on src/main/java/com/dope/breaking/domain/post/Location.java:
```diff
@@ -1,5 +1,7 @@
 package com.dope.breaking.domain.post;
 
+import lombok.Data;
```
삭제 요망 ~

JinuCheon commented on src/main/java/com/dope/breaking/domain/post/Post.java:
```diff
@@ -37,6 +37,8 @@ public class Post {
     @OneToMany(mappedBy = "post")
     private List<Purchase> buyerList = new ArrayList<Purchase>();
 
+    private String title;
```
변경 사항이 있으면, 개별 커밋으로 나누기!

JinuCheon commented on src/main/java/com/dope/breaking/domain/post/Post.java:
```diff
@@ -55,4 +57,20 @@ public class Post {
 
     private LocalDateTime eventTime;
 
+    public void CreatePost(String title, String content, int price){
```
이것은 무엇인가요?

JinuCheon commented on src/main/java/com/dope/breaking/domain/user/User.java:
```diff
@@ -4,6 +4,7 @@
 import com.dope.breaking.domain.financial.Statement;
 import com.dope.breaking.domain.post.Post;
 import lombok.Getter;
+import lombok.NoArgsConstructor;
```
필요없는 패키지는 마지막에 확인 후, 삭제 부탁드려요~~!

JinuCheon commented on src/main/java/com/dope/breaking/dto/user/EmailValidateRequest.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
response dto에서, 파라미터 없는 생성자가 왜 필요한가용?

JinuCheon commented on src/main/java/com/dope/breaking/dto/user/MessageResponse.java:
```diff
@@ -0,0 +1,15 @@
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
+public class MessageResponse {
```
의도가 명확하게, 클래스 명을 바꾸는건 어떨까요??

JinuCheon commented on src/test/java/com/dope/breaking/api/UserAPITest.java:
```diff
@@ -0,0 +1,47 @@
+    @Autowired UserRepository userRepository;
+
+    @Test
+    void signInConfirm() {
```
매서드 명 확인!!

JinuCheon commented on src/main/java/com/dope/breaking/exception/SignInException.java:
```diff
@@ -0,0 +1,24 @@
+package com.dope.breaking.exception;
```
service 계층을 만들 때가 온 것 같습니다..!

JinuCheon commented on src/test/java/com/dope/breaking/exception/SignInExceptionTest.java:
```diff
@@ -0,0 +1,22 @@
+package com.dope.breaking.exception;
```
서비스 계층 만들면, 테스트코드 패키지도 변경 부탁드려요~

JinuCheon reviewed (changes requested):
comment 남긴 부분 확인해서, 수정 부탁드립니다~~!

Martin0o0 commented on build.gradle:
```diff
@@ -1,5 +1,5 @@
 plugins {
-	id 'org.springframework.boot' version '2.7.0'
+	id 'org.springframework.boot' version '2.5.2'
```
버전 확인했습니다. 수정 고맙습니다!

minwukim committed:
chore: update according to the code reviews

- Removed unnecessary packages
- Removed unnecessary annotations
- Relocated isValidEmail function to UserService

minwukim committed:
chore: unify @Getter to @Data for dtos.

Martin0o0 commented on src/main/java/com/dope/breaking/service/UserService.java:
```diff
@@ -0,0 +1,23 @@
+    public static boolean isValidEmail(String email) {
+
+        boolean err = false;
+        String regex = "^[_a-z0-9-]+(.[_a-z0-9-]+)*@(?:\\w+\\.)+\\w+$";
```
이메일을 정규식으로 판별하는 방식 너무 좋은 것 같습니다!

Martin0o0 commented on src/main/java/com/dope/breaking/dto/user/SignUpRequestDto.java:
```diff
@@ -0,0 +1,29 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class SignUpRequestDto {
+
```
추후 RequestDto에 username 필드도 추가 부탁드릴께요!

Martin0o0 commented on src/main/java/com/dope/breaking/api/UserAPI.java:
```diff
@@ -0,0 +1,125 @@
+
+        User user = new User();
+
+        user.signUp(
```
스프링 시큐리티를 위해 추후 username 필드와 password 필드 추가가 필요해 보입니다. 추후 제가 수정해서 PR 날리도록 하겠습니다.

minwukim replied:
현재 수정 중입니다. RequestDto 에도 password가 필요한가요?

Martin0o0 reviewed (approved):
시큐리티 접목하면서 수정사항 몇부분이 필요해 보입니다. 추후 제가 수정하도록 하겠습니다! 그리고 코드 정말 잘 짜신거 같습니다. 수고하셨습니다!

minwukim committed:
feat: update field for "User" entity & SignUpRequestDto

JinuCheon reviewed (approved):
Good 고생하셨어요~
